src/data_gradients/feature_extractors/common/image_duplicates.py

```diff
@@ -0,0 +1,264 @@
+                valid_dups.append([d for d in dup_clique if d.startswith(self.valid_image_dir)])
+
+            # IF THE CLIQUE HAS IT LEAST ONE PATH IN EACH - ADD IT TO intersection_dups
+            if self._is_intersection_dup(dup_clique):
+                intersection_dups.append(dup_clique)
+
+        self.train_dups, self.valid_dups, self.intersection_dups = train_dups, valid_dups, intersection_dups
+        self.train_dups_appearences = self._count_dir_dup_appearences(self.train_dups, self.train_image_dir)
+        self.validation_dups_appearences = self._count_dir_dup_appearences(self.valid_dups, self.valid_image_dir)
+        self.intersection_train_appearnces = self._count_dir_dup_appearences(self.intersection_dups, self.train_image_dir)
+        self.intersection_val_appearnces = self._count_dir_dup_appearences(self.intersection_dups, self.valid_image_dir)
+
+    @staticmethod
+    def _is_in_dup_clique(sample: str, dup_clique_list: List[List[str]]) -> bool:
+        """
+        Whether sample is already in some duplicated clique in dup_clique_list.
+
+        :param sample: str, an image path
+        :param dup_clique_list: List[List[str]], list of duplicate image path cliques (lists).
+        """
+        return any([sample in d for d in dup_clique_list])
+
+    @staticmethod
+    def _make_dup_clique(dup_key, dups):
+        dup_clique = [dup_key] + dups[dup_key]
+        return dup_clique
+
+    def _is_train_dup(self, dup_clique: List[str]) -> bool:
+        """
+        Whether the dup_clique contains duplicate image paths from train_image_dir.
+
+        :param dup_clique: str, list of duplicated image paths
+        """
+        return len([d for d in dup_clique if d.startswith(self.train_image_dir)]) > 1
+
+    def _is_valid_dup(self, dup_clique: List[str]) -> bool:
+        """
+        Whether the dup_clique contains duplicate image paths from valid_image_dir.
+
+        :param dup_clique: str, list of duplicated image paths
+        """
+        return len([d for d in dup_clique if d.startswith(self.valid_image_dir)]) > 1
+
+    def _is_intersection_dup(self, dup_clique: List[str]) -> bool:
+        """
+        Whether the dup_clique contains duplicate image paths from train_image_dir and valid_image_dir.
+
+        :param dup_clique: str, list of duplicated image paths
+        """
+        return len([d for d in dup_clique if d.startswith(self.train_image_dir)]) > 0 and len([d for d in dup_clique if d.startswith(self.valid_image_dir)]) > 0
+
+    @staticmethod
+    def _count_dup_appearences(dups: List[List[str]]) -> int:
+        """
+        Counts the duplicate appearences = sum of the sizes of all duplicate cliques in dups.
+        """
+        return sum([len(d) for d in dups])
+
+    def _count_dir_dup_appearences(self, dups: List[List[str]], dir: str) -> int:
+        """
+        Counts the duplicate appearences = sum of the sizes of all duplicate cliques in dups.
```

Comment:
Same description as `_count_dup_appearences`, what is the difference? `_count_dir_dup_appearences` filters the paths to only keep these listed in `dir` ?

Reply:
Yup, fixed docs in latest commit.